### PR TITLE
Update model signature inputs to be optional

### PR DESCRIPTION
**Pull request by serena-ruan**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?
Make ModelSignature inputs an optional field, so that in some cases we could avoid input schema enforcement by setting inputs schema to None.

## How is this patch tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/models/signature.py [hunk 1]
```diff
@@ -103,16 +108,19 @@ def from_dict(cls, signature_dict: Dict[str, Any]):
...
-        if "outputs" in signature_dict and signature_dict["outputs"] is not None:
-            outputs = Schema.from_json(signature_dict["outputs"])
-        else:
-            outputs = None
-        if (params := signature_dict.get("params")) is not None:
-            params = ParamSchema.from_json(params)
-            return cls(inputs, outputs, params)
-        else:
-            return cls(inputs, outputs)
+        inputs = (
```
Comment (BenWilson2) [inline mlflow/models/signature.py:111]: awesome refactoring :)

**Comment by serena-ruan** [comment]
Verified locally:
Model signature only contains output:
image
image

Model signature only contains params:
image
image

File: mlflow/models/signature.py [hunk 2]
```diff
@@ -70,6 +73,8 @@ def __init__(self, inputs: Schema, outputs: Schema = None, params: ParamSchema =
                 "If params are provided, they must by of type mlflow.models.signature.ParamSchema. "
                 f"Got '{type(params)}'"
             )
+        if all(x is None for x in [inputs, outputs, params]):
```
Comment (harupy) [inline mlflow/models/signature.py:76]: /ai can we use `[inputs, outputs, params]`.count(None) here? Is it faster or slower?

File: mlflow/models/signature.py [hunk 3]
```diff
@@ -103,16 +108,19 @@ def from_dict(cls, signature_dict: Dict[str, Any]):
...
-            params = ParamSchema.from_json(params)
-            return cls(inputs, outputs, params)
-        else:
-            return cls(inputs, outputs)
+        inputs = (
+            Schema.from_json(signature_dict["inputs"]) if signature_dict.get("inputs") else None
+        )
+        outputs = (
+            Schema.from_json(signature_dict["outputs"]) if signature_dict.get("outputs") else None
+        )
```
Comment (harupy) [inline mlflow/models/signature.py:116]: /ai can this be simplified with assignment expression?

Comment (harupy) [reply mlflow/models/signature.py:116]: /ai Thanks! Does the followign code work?

```python
Schema.from_json(x) if x := signature_dict.get("outputs") else None
```

Comment (harupy) [reply mlflow/models/signature.py:116]: /ai do you think this is readable?

File: mlflow/models/signature.py (hunk 2 above)
Comment (harupy) [reply mlflow/models/signature.py:76]: /ai Thanks! How would you refactor this line?

File: mlflow/models/signature.py [hunk 4]
```diff
@@ -57,10 +57,13 @@ class ModelSignature:
...
 
-    def __init__(self, inputs: Schema, outputs: Schema = None, params: ParamSchema = None):
-        if not isinstance(inputs, Schema):
-            raise TypeError(f"inputs must be mlflow.models.signature.Schema, got '{type(inputs)}'")
-        if outputs is not None and not isinstance(outputs, Schema):
+    def __init__(self, inputs: Schema = None, outputs: Schema = None, params: ParamSchema = None):
+        if inputs and not isinstance(inputs, Schema):
+            raise TypeError(
+                "inputs must be either None or mlflow.models.signature.Schema, "
+                f"got '{type(inputs)}'"
```
Comment (harupy) [inline mlflow/models/signature.py:64]:
```suggestion
                f"got '{type(inputs).__name__}'"
```

**Review by harupy — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/9450/files#r1305306697 is addresed!

**Comment by BenWilson2** [comment]
Could we add the integration test to be extra certain for future changes? Something similar to this?
Perhaps in the ``test_cli.py`` module?

```python
@pytest.mark.parametrize(
    ["input_schema", "output_schema", "params_schema"],
    [(True, False, False), (False, True, False), (False, False, True)],
)
def test_signature_enforcement_with_model_serving(input_schema, output_schema, params_schema):
    class MyModel(pyfunc.PythonModel):
        def predict(self, context, model_input, params=None):
            return ["test"]

    def setup_signature(input_schema, output_schema, params_schema):
        """Setup signature based on input, output and params schema."""
        input_data = ["testing"] if input_schema else None
        output_data = ["test"] if output_schema else None
        params = {"test": "test"} if params_schema else None

        return infer_signature(model_input=input_data, model_output=output_data, params=params)

    input_data = ["testing"]
    signature = setup_signature(input_schema, output_schema, params_schema)

    with mlflow.start_run():
        model_info = mlflow.pyfunc.log_model(
            artifact_path="signature", python_model=MyModel(), signature=signature
        )

    inference_payload = json.dumps({"inputs": input_data})

    # Serve and score the model
    scoring_result = pyfunc_serve_and_score_model(
        model_uri=model_info.model_uri,
        data=inference_payload,
        content_type=pyfunc.scoring_server.CONTENT_TYPE_JSON,
        extra_args=["--env-manager", "local"],
    )
    scoring_result.raise_for_status()

    # Assert the prediction result
    assert json.loads(scoring_result.content)["predictions"] == ["test"]
 ```

**Comment by serena-ruan** [comment]
Verified in test shard that this makes feature store models work
image

**Review by BenWilson2 — APPROVED** [review]
LGTM! Thank you for the fix @serena-ruan

**Pull request merged**